Add  Traversing instance for Cell

miguel-negrao opened the pull request:
Hopefully this solves #56 and #57 

With this instance should `resampleList` and `resampleMaybe` be removed ?

```
> fst <$> steps (traverse' sumC :: Cell IO [Int] [Int]) [[1,2], [1],[3,4]] 
[[0,1],[3],[4,7]]
> fst <$> steps (traverse' sumC :: Cell IO (Maybe Int) (Maybe Int)) [Just 1, Nothing, Nothing, Just 2, Just 3, Nothing] 
[Just 0,Nothing,Nothing,Just 1,Just 3,Nothing]
```

miguel-negrao commented:
Before the last commit my version of traverse' was not as strict as `resampleList`:

```
> take 4 $ head $ fst $ runIdentity $ steps (traverse' (arr id) :: Cell Identity [Int] [Int]) [[1..],[2],[3]]
[1,2,3,4]
> take 4 $ head $ fst $ runIdentity $ steps (resampleList (arr id) :: Cell Identity [Int] [Int]) [[1..],[2],[3
...
```

After the last commit both of the expressions above result in bottom.

miguel-negrao commented:
Refactored to its own module `LiveCoding.Profunctors` and added tests.

turion commented on essence-of-live-coding/src/LiveCoding/Cell/Profunctor.hs:
```diff
@@ -0,0 +1,81 @@
+instance Monad m => Traversing (Cell m) where
+  traverse' (Cell state step) = Cell state step' where
+    step' s a = runStateT (traverse (\a -> StateT (`step` a)) a) s
+  traverse' (ArrM f) = ArrM (traverseStrict f)
```
The `StrictT` trick is very interesting, but I have a bad feeling to introduce so much generic code which should reside in a separate package, just for this one special case.

I don't understand why `traverseStrict` is needed here, and not in the `Cell` case. And maybe it's possible to implement `traverseStrict` in a simpler way?

Also, I don't understand what exactly is strict here. The container structure we're traversing over? Or the individual values? Because the latter doesn't make sense to me (even though I implemented `resampleList` that way).

turion replied:
> I don't understand why `traverseStrict` is needed here, and not in the `Cell` case. And maybe it's possible to implement `traverseStrict` in a simpler way?

Ah, you said that it's because of strict state! Makes sense.

turion commented:
I'm unsure about the strictness. I agree that there should be some strictness, but I think the examples shouldn't be so strict. In fact, I'm more and more convinced that I implemented `resampleList` incorrectly.

Some code examples:

```haskell

testSetup x = take 4 $ head $ fst $ runIdentity $ steps (traverse' (arr id) :: Cell Identity [Int] [Int]) x

shouldNotBeBottom = testSetup [[1..],[2],[3]]

shouldBeBottom = testSetup [1 : error "boom", [2], [3]]
...

I think it should only diverge if there is a a problem in the list, but not if there is a problem in the data. That way we can be sure the list is processed completely, but leave it to the cell itself to decide whether it needs to evaluate all the data inside.

Can you add a few tests that capture the expected behaviour? If you use `error` or `undefined` instead of an infinite list, you can use [`assertThrows` or exception handling](https://stackoverflow.com/questions/46330592/is-it-possible-to-assert-an-error-case-in-hunit) to make sure it explodes in the right cases.

turion commented on essence-of-live-coding/src/LiveCoding/Cell/Profunctor.hs:
```diff
@@ -0,0 +1,81 @@
+-- profunctors
+import Data.Profunctor
+import Data.Profunctor
+    ( Strong, WrappedArrow(WrapArrow), Profunctor )
```
```suggestion
    ( Strong, WrappedArrow, Profunctor )
```

turion replied:
That line is superseded by the next one I guess?

turion commented on essence-of-live-coding/src/LiveCoding/Cell/Profunctor.hs:
```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE DeriveAnyClass #-}
+
+-- export only instances
```
Ok, but we should export them always. I think it's best to avoid the orphan instance and add this instance to `Cell.lhs`.

turion replied:
I mean, you've added it to `LiveCoding.hs`, which is great for users, but it also means that we can't import it internally in the library (because that would be a circular import).

miguel-negrao replied:
So I should move all the code to `Cell.lhs`, correct ?

turion replied:
Yes.

turion commented on essence-of-live-coding/src/LiveCoding/Cell/Profunctor.hs:
```diff
@@ -0,0 +1,81 @@
+
+deriving via (WrappedArrow (Cell m)) instance Monad m => Profunctor (Cell m) 
+deriving via (WrappedArrow (Cell m)) instance Monad m => Strong (Cell m)
+deriving via (WrappedArrow (Cell m)) instance Monad m => Data.Profunctor.Choice.Choice (Cell m) 
```
```suggestion
deriving via (WrappedArrow (Cell m)) instance Monad m => Choice (Cell m) 
```

turion commented:
> Hopefully this solves #56 and #57

Yes, it's looking that way!

> With this instance should `resampleList` and `resampleMaybe` be removed ?

No, I think they should be replaced with type-specific aliases `resampleMaybe = traverse'`. The same with `resample`, which works over a `Vector`.

miguel-negrao commented:
> I'm more and more convinced that I implemented `resampleList` incorrectly.

I had implemented this way to match the strictness in `resampleList`, so that changing to an implementation of `resampleList` based on traverse' would keep the same behaviour.

> I think it should only diverge if there is a a problem in the list, but not if there is a problem in the data. That way we can be sure the list is processed completely, but leave it to the cell itself to decide whether it needs to evaluate all the data inside.

Ok, in that case I will change to using `traverse` and lazy `StateT`.

turion commented:
> Ok, in that case I will change to using `traverse` and lazy `StateT`.

Will that evaluate the list structure strictly? I'd expect it doesn't.

turion commented:
Some background: I introduced the additional strictness in https://github.com/turion/essence-of-live-coding/pull/19. In fact, we have strictness in the data passing through cells:

```haskell
  Cell state2 step2 . Cell state1 step1 = Cell { .. }
    where
      cellState = Composition state1 state2
      cellStep (Composition state1 state2) a = do
        (!b, state1') <- step1 state1 a
        (!c, state2') <- step2 state2 b
        return (c, Composition state1' state2')
```

This surprises me a lot. I think this should be rethought first.

miguel-negrao commented:
> Will that evaluate the list structure strictly? I'd expect it doesn't.

The traversal of the traversable structure is lazy when using State.Lazy:

```Haskell
Prelude Control.Monad.Trans.State.Lazy> take 2 $ fst $ flip runState 0 $ traverse (\_ -> modify (+2) >> get) [1..]
[2,4]
```

On the other hand from the code of the composition of Cell above it is probably going to evaluate the whole list before passing it on to the next cell. So the behavior might be different if we just test a single cell or if we test the composition of two cells (i.e. `cell` vs `cell >> (arr id)` ). I will create some tests for this, so we can check on the behavior. 

But I suspect it won't matter much if we use Strict or Lazy State when the cell is composition of two cells...

miguel-negrao commented:
I've refactored the PR:


- Renamed `Choice` to `CellChoice` so it doesn't clash with profunctors' `Choice`.
- Moved code to `Cell.lhs`
- `traverse'` is using `Control.Monad.Trans.State.Lazy`.
- Reimplemented resampleList, etc as just `traverse'`.
- Added tests for `traverse'`'s lazyness.

This implementation keeps the list created in `resampleList` lazy. This is the case even when composing with another Cell. I suppose that is the case because the bang pattern used in `cell1 >>> cell2` on a list will evaluate the list into either `Nil` or `_ : _ `, which doesn't force the entire list, right ?

Check if this addresses all the issues.

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,49 @@ import LiveCoding
   [ testProperty "steps produces outputs"
     $ \(inputs :: [Int]) -> inputs === fst (runIdentity $ steps (id :: Cell Identity Int Int) inputs)
+  , testProperty "traverse' works as expected for Lists"
+  $ \(inputs :: [[Int]]) -> fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
```
```suggestion
    $ \(inputs :: [[Int]]) -> fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
```

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,49 @@ import LiveCoding
+  , testProperty "traverse' works as expected for Maybes"
+  $ \(inputs :: [Maybe Int]) -> fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
+  , testProperty "traverse' for [] is equal to resampleList"
+  $ \(inputs :: [[Int]]) -> fst (runIdentity $ steps (resampleList (arr (+10) :: Cell Identity Int Int)) inputs)
+    === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
```
```suggestion
    $ \(inputs :: [[Int]]) -> fst (runIdentity $ steps (resampleList (arr (+10) :: Cell Identity Int Int)) inputs)
      === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs) 
```

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,49 @@ import LiveCoding
+  , testProperty "traverse' for Maybe is equal to resampleMaybe"
+  $ \(inputs :: [Maybe Int]) -> fst (runIdentity $ steps (resampleMaybe (arr (+10) :: Cell Identity Int Int)) inputs)
+    === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
+  , testCase "traverse' by itself does not force the entire list (ArrM)" $ 
+    testError "Traverse' is too strict" id $ 
+    head $ head $ fst $ runIdentity $ steps (traverse' (arr id) :: Cell Identity [Int] [Int]) [[1, error "Bang !"]]
```
```suggestion
  , testCase "traverse' by itself does not force the entire list (ArrM)" 
    $ testError "Traverse' is too strict" id
    $ head $ head $ fst $ runIdentity $ steps (traverse' (arr id) :: Cell Identity [Int] [Int]) [[1, error "Bang !"]]
```
And similarly the other cases

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,49 @@ import LiveCoding
+  $ \(inputs :: [[Int]]) -> fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
+  , testProperty "traverse' works as expected for Maybes"
+  $ \(inputs :: [Maybe Int]) -> fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
+  , testProperty "traverse' for [] is equal to resampleList"
```
This is given by definition ;)

miguel-negrao replied:
Ah, yes, this test was for when resampleList was not just traverse'. Removed these two tests.

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,49 @@ import LiveCoding
 test = testGroup "Cell"
   [ testProperty "steps produces outputs"
     $ \(inputs :: [Int]) -> inputs === fst (runIdentity $ steps (id :: Cell Identity Int Int) inputs)
+  , testProperty "traverse' works as expected for Lists"
```
What's the law you're expecting to hold?

miguel-negrao replied:
I was checking whether traverse' applies the function to each element of the list.

turion replied:
So something like `traverse' (cell >>> arr f) = traverse' cell >>> arr (map f)`?

turion commented:
>     * Renamed `Choice` to `CellChoice` so it doesn't clash with profunctors' `Choice`.

I'd prefer it if `Choice` from `profunctor` were aliased. Does that work as well?

>     * `traverse'` is using `Control.Monad.Trans.State.Lazy`.
>     * Added tests for `traverse'`'s lazyness.
> 
> This implementation keeps the list created in `resampleList` lazy. This is the case even when composing with another Cell. I suppose that is the case because the bang pattern used in `cell1 >>> cell2` on a list will evaluate the list into either `Nil` or `_ : _ `, which doesn't force the entire list, right ?

So an input of `"ok" : error "oh noez"` would not crash? I guess you're right, we don't seq the whole list structure, but only the first constructor when we do the bang pattern on the value passing through the cell.

Can you check whether the `SpeedTest` in `essence-of-live-coding-yampa-speedtest` still doesn't have a space leak? This used to be one of the reasons for the strictness change. (I tried to check with memory profiling, but I couldn't do it.)

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,49 @@ import LiveCoding
+    head $ head $ fst $ runIdentity $ steps (traverse' cellId :: Cell Identity [Int] [Int]) [[1, error "Bang !"]]
+  , testCase "traverse' of a Cell composed with another Cell does not force the entire list" $ 
+    testError "Traverse' is too strict" id $ 
+    head $ head $ fst $ runIdentity $ steps (traverse' cellId >>> cellId  :: Cell Identity [Int] [Int]) [[1, error "Bang !"]]
```
I think you can refactor the tests with `CellSimulation`, they will be easier to understand afterwards :)

miguel-negrao replied:
But here I'm not checking whether the cell produces a specific value, but just that it doesn't produce an error, so I don't think I can use `CellSimulation` for that, can I ?

miguel-negrao commented:
> > ```
> > * Renamed `Choice` to `CellChoice` so it doesn't clash with profunctors' `Choice`.
> > ```
> 
> I'd prefer it if `Choice` from `profunctor` were aliased. Does that work as well?

Sure, no problem, will do that :). If the user needs to use profunctor's Choice then they must import qualified by themselves, and profunctor's Choice probably is not going to be used.

> > This implementation keeps the list created in `resampleList` lazy. This is the case even when composing with another Cell. I suppose that is the case because the bang pattern used in `cell1 >>> cell2` on a list will evaluate the list into either `Nil` or `_ : _ `, which doesn't force the entire list, right ?
> 
> So an input of `"ok" : error "oh noez"` would not crash? I guess you're right, we don't seq the whole list structure, but only the first constructor when we do the bang pattern on the value passing through the cell.
> 
Yes, it doesn't crash. The tests that I added are testing this.

```Haskell
  , testCase "traverse' by itself does not force the entire list (Cell)" $ 
    testError "Traverse' is too strict" id $ 
    head $ head $ fst $ runIdentity $ steps (traverse' cellId :: Cell Identity [Int] [Int]) [[1, error "Bang !"]]
```

> Can you check whether the `SpeedTest` in `essence-of-live-coding-yampa-speedtest` still doesn't have a space leak? This used to be one of the reasons for the strictness change. (I tried to check with memory profiling, but I couldn't do it.)

I profiled SpeedTest and I didn't see any space leak. The output has a NaN, is that normal ? Looking at the code of SpeedTest, I think that it doesn't use any resample function, so there's no reason why it should change behavior right ?

```
essence-of-live-coding$ stack exec --profile -- SpeedTest +RTS -p -hc
(NaN,NaN,-5.56184561945215e306)
essence-of-live-coding$ stack exec -- hp2ps -e8in -c SpeedTest.hp
essence-of-live-coding$ xdg-open SpeedTest.ps
```

![Captura de ecrã de 2022-03-05 18-22-57](https://user-images.githubusercontent.com/382777/156895595-7c6e4cb1-f409-459f-9f4f-dc334e692c2b.png)

turion commented:
> > > This implementation keeps the list created in `resampleList` lazy. This is the case even when composing with another Cell. I suppose that is the case because the bang pattern used in `cell1 >>> cell2` on a list will evaluate the list into either `Nil` or `_ : _ `, which doesn't force the entire list, right ?
> > 
> > 
> > So an input of `"ok" : error "oh noez"` would not crash? I guess you're right, we don't seq the whole list structure, but only the first constructor when we do the bang pattern on the value passing through the cell.
> 
> Yes, it doesn't crash. The tests that I added are testing this.
> 
> ```haskell
>   , testCase "traverse' by itself does not force the entire list (Cell)" $ 
>     testError "Traverse' is too strict" id $ 
>     head $ head $ fst $ runIdentity $ steps (traverse' cellId :: Cell Identity [Int] [Int]) [[1, error "Bang !"]]
> ```

I think that's not the same thing what I meant. The test I meant would have as input not `[[1, error "Bang !"]]`, but `[1 : error "Bang !"]`. This is subtly different.

> > Can you check whether the `SpeedTest` in `essence-of-live-coding-yampa-speedtest` still doesn't have a space leak? This used to be one of the reasons for the strictness change. (I tried to check with memory profiling, but I couldn't do it.)
> 
> I profiled SpeedTest and I didn't see any space leak.

Awesome, thanks :)

> The output has a NaN, is that normal ?

Yes, I think at some point it exceeds the limits of a `Double`.

> Looking at the code of SpeedTest, I think that it doesn't use any resample function, so there's no reason why it should change behavior right ?

True, I'm puzzled a bit. I thought it used it somewhere, but maybe that was an older temporary version of the test.

Or maybe I simply changed the definition of `>>>`, and thought that I'd have to change the definition of `resampleList` analogously for consistency.

Anyways. Your choice of lazyness seems to be a good one then :) let's keep it like it is now.

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,45 @@ import LiveCoding
+  ]  
+
+testError :: String -> (Bool -> Bool) -> a -> IO ()
+testError message alterBoolValue value = do
```
`alterBoolValue` is `id` in all usage cases, it seems. Can you remove it?

miguel-negrao replied:
This function is no longer needed.

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,45 @@ import LiveCoding
+    handler :: ErrorCall -> IO Bool
+    handler e = pure True
+  errored <- catch (pure $ seq value False) handler
+  if alterBoolValue errored then
+    assertFailure message
+  else
+    pure ()
```
Can you use `when`?

miguel-negrao replied:
yup

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,45 @@ import LiveCoding
-  ]
+  ]  
+
+testError :: String -> (Bool -> Bool) -> a -> IO ()
```
```suggestion
-- tests whether the value `a` throws an exception
testError :: String -> (Bool -> Bool) -> a -> IO ()
```

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,45 @@ import LiveCoding
+  let 
+    handler :: ErrorCall -> IO Bool
+    handler e = pure True
+  errored <- catch (pure $ seq value False) handler
```
I'm confused. Will this be `True` when there was an exception? I believe not.

miguel-negrao replied:
Yes, I believe so. If there is no exception in value then `seq value False` returns `False`. If there is an exception then handler will run which will return `True`.

turion replied:
But that means that the test _fails_ if there is an exception, right? For that use case, no extra function is needed. If an error is raised anywhere in a test, that counts as a test failure. What I wanted was a function where the test _succeeds_ if there is an exception in the value. That way you can test that a function has a certain strictness, by storing an exception somewhere in the place where you expect strictness, and making sure the whole thing explodes. But I think we don't have these places anymore, because `traverse'` is now lazy.

Maybe it's easiest to drop the tests with the errors, since we don't guarantee strictness or laziness anyways.

miguel-negrao replied:
> But that means that the test fails if there is an exception, right? 
yes.

> For that use case, no extra function is needed. If an error is raised anywhere in a test, that counts as a test failure.

Ah, I was missing that bit. Then I could change the tests to just a normal CellSimulation. Should I do that, or just drop the tests ?

>  What I wanted was a function where the test succeeds if there is an exception in the value. 

That is just a question of changing the  `errored` to `not errored` in the function `testError` , then it will fail the test if there is no error.

turion replied:
> > For that use case, no extra function is needed. If an error is raised anywhere in a test, that counts as a test failure.
> 
> Ah, I was missing that bit. Then I could change the tests to just a normal CellSimulation. Should I do that, or just drop the tests ?
> 
> > What I wanted was a function where the test succeeds if there is an exception in the value.
> 
> That is just a question of changing the `errored` to `not errored` in the function `testError` , then it will fail the test if there is no error.

I think you can either drop the tests or change them to a `CellSimulation` if you want.

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -117,4 +117,14 @@ test = testGroup "Utility unit tests"
     , output1 = [Nothing, Just 2]
     , output2 = [Just 3, Just 4]
     }
+    , testProperty "traverse' works as expected for Lists"
+    $ \(inputs :: [[Int]]) -> fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
+    , testProperty "traverse' works as expected for Maybes"
+    $ \(inputs :: [Maybe Int]) -> fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
+    , testProperty "traverse' for [] is equal to resampleList"
+    $ \(inputs :: [[Int]]) -> fst (runIdentity $ steps (resampleList (arr (+10) :: Cell Identity Int Int)) inputs)
+      === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
+    , testProperty "traverse' for Maybe is equal to resampleMaybe"
+    $ \(inputs :: [Maybe Int]) -> fst (runIdentity $ steps (resampleMaybe (arr (+10) :: Cell Identity Int Int)) inputs)
+      === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
   ]
```
But these can be written with `CellSimulation`, right?

miguel-negrao replied:
indeed

turion commented on essence-of-live-coding/test/Cell.hs:
```diff
@@ -23,10 +25,45 @@ import LiveCoding
 test = testGroup "Cell"
   [ testProperty "steps produces outputs"
     $ \(inputs :: [Int]) -> inputs === fst (runIdentity $ steps (id :: Cell Identity Int Int) inputs)
+  , testProperty "traverse' works as expected for Lists"
+    $ \(inputs :: [[Int]]) -> 
+      fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
+  , testProperty "traverse' works as expected for Maybes"
+    $ \(inputs :: [Maybe Int]) -> 
+      fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
```
Are these tests duplicate to the ones in `test/Cell/Util.hs`?

I think all tests regarding `traverse'` should be in `test/Cell/Util.hs`, and ideally in a separate `testGroup "traverse'"`.

(I know not all tests are sorted like that currently, but they should be ideally.)

miguel-negrao replied:
Ah, yes, has forgotten to delete them in Util.hs, will move tests to Util.hs

miguel-negrao commented:
> I think that's not the same thing what I meant. The test I meant would have as input not [[1, error "Bang !"]], but [1 : error "Bang !"]. This is subtly different.

Ah, that is quite subtly different indeed. So one is e 1 : _|_ : Nil and the other 1 : _|_ , right ?

In any case, it also doesn't cause an error. I will change the tests to use 1 : _|_ instead of 1 : _|_ : Nil.

miguel-negrao commented:
Ok, I squashed the PR as it was too complicated solving conflicts with master otherwise.

Added CellIdentitySimulation to test whether one cell behaves the same as another cell.
Added tests for law `traverse' (arr f) = arr (f <$>)`
Added tests for traverse' with arbitrary Cells created with random state functions (s -> a -> (b,s)) which don't operate on the monad. It's pretty awesome that quickcheck can create random functions !

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -89,9 +103,9 @@ test = testGroup "Utility unit tests"
-        fst (runIdentity $ steps (delay startValue >>> changes >>> hold startValue) inputs)
+  , testProperty "delay a >>> changes >>> hold a = delay a"
+    $ \(inputs :: [Int]) (startValue :: Int) ->
+      CellIdentitySimulation  (delay startValue) (delay startValue >>> changes >>> hold startValue) inputs
```
```suggestion
      CellIdentitySimulation (delay startValue) (delay startValue >>> changes >>> hold startValue) inputs
```

miguel-negrao replied:
done

turion commented on essence-of-live-coding/test/Util.hs:
```diff
@@ -81,3 +81,18 @@ instance (Eq b, Show b) => Testable (CellSimulation a b) where
+-- | Basic unit test for 'Cell' identities.
+--   Check whether one cell behaves the same as another cell.
+data CellIdentitySimulation a b = CellIdentitySimulation
+  { cella :: Cell Identity a b
+  , cellb :: Cell Identity a b
+  , commonInput :: [a]  
```
```suggestion
  { cell1 :: Cell Identity a b
  , cell2 :: Cell Identity a b
  , input :: [a]  
```
So it's more like the other types, and there is no confusion with `a`/`b`

miguel-negrao replied:
I was actually trying to avoid the names used in the other CellMigrationSimulation, unfortunately this is the dreaded issue of these functions ending up in a common namespace. Is there way around the record issue ?

turion replied:
Ah ok I see. I typically don't mind because I rarely use the record functions. I just like record syntax & wildcards, and there it's no trouble if the names are duplicate.

miguel-negrao replied:
We can't use the same names, so I changed them to cell1', cell2' and input'.

turion commented on essence-of-live-coding/test/Util.hs:
```diff
@@ -81,3 +81,18 @@ instance (Eq b, Show b) => Testable (CellSimulation a b) where
     , output2 = []
     }
+
+-- | Basic unit test for 'Cell' identities.
```
This is an awesome idea! Thanks :)

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -150,3 +164,58 @@ test = testGroup "Utility unit tests"
+  -> CellSimulation (t a) (t b)
+traverseCellTest s f input = CellSimulation
+  { cell = traverse' (Cell s (\s a -> pure $ f s a)) :: Cell Identity (t a) (t b)
+  ,  output = runIdentity $ evalStateT (traverse (traverse (\a -> StateT (Identity . (`f` a)) )) input) s
```
```suggestion
  , output = runIdentity $ evalStateT (traverse (traverse (\a -> StateT (Identity . (`f` a)) )) input) s
```

miguel-negrao replied:
done

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -150,3 +164,58 @@ test = testGroup "Utility unit tests"
+
+testTraverse' :: Test
+testTraverse' = testGroup "Traversing unit tests"
+  [ testProperty "traverse' (arr f) = arr (f <$>) ([a])"
+    $ \(input :: [[Int]], Fn (f :: Int -> Int)) -> traverseArrLaw f input
+  , testProperty "traverse' (arr f) = arr (f <$>) (Maybe a)"
+    $ \(input :: [Maybe Int], Fn (f :: Int -> Int)) -> traverseArrLaw f input
+  , testProperty "traverse' (arr f) = arr (f <$>) (Vecttor 4 Int)"
+    $ \(input :: [(Int,Int,Int,Int)], Fn (f :: Int -> Int)) -> traverseArrLaw f (V.fromTuple <$> input)
```
You can put these under a common `testGroup`, similarly the other tests.

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -150,3 +164,58 @@ test = testGroup "Utility unit tests"
+    $ \(input :: [[Int]], Fn (f :: Int -> Int)) -> traverseArrLaw f input
+  , testProperty "traverse' (arr f) = arr (f <$>) (Maybe a)"
+    $ \(input :: [Maybe Int], Fn (f :: Int -> Int)) -> traverseArrLaw f input
+  , testProperty "traverse' (arr f) = arr (f <$>) (Vecttor 4 Int)"
```
```suggestion
  , testProperty "traverse' (arr f) = arr (f <$>) (Vector 4 Int)"
```

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -150,3 +164,58 @@ test = testGroup "Utility unit tests"
+testTraverse' :: Test
+testTraverse' = testGroup "Traversing unit tests"
+  [ testProperty "traverse' (arr f) = arr (f <$>) ([a])"
+    $ \(input :: [[Int]], Fn (f :: Int -> Int)) -> traverseArrLaw f input
```
Instead of replicating the same test for several types, maybe you can also create types

```haskell
data TestInput = IntList [Int] | MaybeInt (Maybe Int) | VectorOfInts (Int, Int, Int, Int)
  deriving (Eq, Show, Arbitrary, CoArbitrary)
```
And then use that type for `input` and `f`.

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -8,9 +10,15 @@ import Data.Functor.Identity
 -- transformers
 import Control.Monad.Trans.Reader
+import Control.Monad.Trans.State.Lazy
+
+
+-- vector-sized
```
```suggestion

-- vector-sized
```

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -150,3 +164,58 @@ test = testGroup "Utility unit tests"
+  }
+
+-- | Tests whether the value `a` throws an exception.
+testError :: String  -> a -> IO ()
```
```suggestion
testError :: String -> a -> IO ()
```

miguel-negrao replied:
Function was removed.

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -150,3 +164,58 @@ test = testGroup "Utility unit tests"
+    $ testError "Traverse' is too strict"
+    $ head $ head $ fst $ runIdentity $ steps (traverse' cellId :: Cell Identity [Int] [Int]) [1 : error "Bang !"]
+  , testCase "traverse' of a Cell composed with another Cell does not force the entire list"
+    $ testError "Traverse' is too strict"
```
Ok, so this tests that in fact this throws an exception, or not?

miguel-negrao replied:
The test succeeds if it does not throw an exception. So in this case as you mentioned elsewhere it is simpler to just test with CellSimulation, since if it is too strict then it will throw an exception and that will consider the test as failed.

turion commented on essence-of-live-coding/test/Util.hs:
```diff
@@ -81,3 +81,18 @@ instance (Eq b, Show b) => Testable (CellSimulation a b) where
+
+instance (Eq b, Show b) => Testable (CellIdentitySimulation a b) where
+  property CellIdentitySimulation { .. }
+    = let 
```
Even simpler you can do something like

```haskell
= property CellSimulation { input1 = input, input2 = input, .. }
```

miguel-negrao replied:
But CellSimulation only has cell, input and output . I don't see how the instance can be created from either CellSimulation or CellMigrationSimulation, can it ?

miguel-negrao commented:
I've added a mechanism to succinctly generate tests for all Traversable instances of interest. It uses induction via type-classes to go through a type-level list a generate a test for each type. If there are any other Traversables that we are interested in, I can add them to the list. Also, I rebased on top of master.

issues:
- I'm getting a warning from haskell language server which I think is bogus:
```
Pattern match(es) are non-exhaustive
In an equation for ‘traverseCellTest’:
    Patterns not matched:
        Proxy _ (Fun _ _) []
        Proxy _ (Fun _ _) (_:_)compile(-Wincomplete-patterns)
```
related to `traverseCellTest _ s (Fn2 f) input = CellSimulation ...` where I'm using the view pattern `Fn`.
- I tried using less Proxies and more type application but I don't know how to tell GHC that a type `a` inside has free type variables, i.e., `a` instead of `Proxy t -> a`. Anyway, it works fine with `Proxy`s.

miguel-negrao commented:
Since more PRs area appearing, it would be nice to be able to finish this one :-). Our last discussion was about the tests.

miguel-negrao commented:
> 

Ah, sorry, I think I missed something :-). which changes are you referring to ?

turion commented:
If you scroll up a bit, you should see a lot of pending discussions :wink: Mostly of stylistic nature.

miguel-negrao commented:
Hi @turion :-). I went through all the discussions above again, there were one or two which were still marked unresolved, and I changed them now to resolved, but in any case I believe my last push from March had hopefully already addressed all the issues (at least that was my intention :-) ). In any case, I might have missed something, off course.

turion commented on essence-of-live-coding/src/LiveCoding/Cell.lhs:
```diff
@@ -412,6 +421,15 @@ Then we can execute the live program in the same way as before.
+deriving via (WrappedArrow (Cell m)) instance Monad m => Data.Profunctor.Choice.Choice (Cell m)
+
+instance Monad m => Traversing (Cell m) where
+  traverse' (Cell state step) = Cell state step' where
+    step' s a = runStateT (traverse (\a -> StateT (`step` a)) a) s
```
```suggestion
  traverse' Cell { cellState, cellStep } = Cell
    { cellStep = \s a -> runStateT (traverse (\a -> StateT (`step` a)) a) s
    , ..
    }
```

miguel-negrao replied:
If I apply that change I get the following error:

```
src/LiveCoding/Cell.lhs:431:7: error:
    • Couldn't match expected type ‘Cell m a b’ with actual type ‘s’
      ‘s’ is a rigid type variable bound by
        a pattern with constructor:
          Cell :: forall (m :: * -> *) a b s.
                  Data s =>
                  s -> (s -> a -> m (b, s)) -> Cell m a b,
        in an equation for ‘traverse'’
        at src/LiveCoding/Cell.lhs:429:13-40
    • In the ‘cellState’ field of a record
      In the expression:
        Cell
          {cellStep = \ s a
                        -> runStateT (traverse (\ a -> StateT (`step` a)) a) s,
           ..}
      In an equation for ‘traverse'’:
          traverse' Cell {cellState, cellStep}
            = Cell
                {cellStep = \ s a
                              -> runStateT (traverse (\ a -> StateT (`step` a)) a) s,
                 ..}
    • Relevant bindings include
        cellStep :: s -> a -> m (b, s)
          (bound at src/LiveCoding/Cell.lhs:429:31)
        cellState :: s (bound at src/LiveCoding/Cell.lhs:429:20)
        traverse' :: Cell m a b -> Cell m (f a) (f b)
          (bound at src/LiveCoding/Cell.lhs:429:3)
    |
431 |     , ..
    |       ^^
```

miguel-negrao replied:
I don't see a way to get rid of this error, so I think we can leave it as it is ?

turion reviewed (changes requested):
Ahhh, my bad! I didn't click on "Submit review"! So only i saw my missing comments. Sorry for delaying this for so long.

miguel-negrao commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -168,4 +196,75 @@ test = testGroup "Utility unit tests"
+genTraversableTests' message f = testGroup message $ genTraversableTests (f message) (Proxy :: Proxy a)
+
+instance (Arbitrary a, KnownNat n) => Arbitrary (V.Vector n a) where
+  arbitrary = sequence $ V.replicate arbitrary
```
Yes, I think mainly the shrink was missing. Will add it.

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -168,4 +196,75 @@ test = testGroup "Utility unit tests"
+instance (Arbitrary a, KnownNat n) => Arbitrary (V.Vector n a) where
+  arbitrary = sequence $ V.replicate arbitrary
+
+cellId :: Monad m => Cell m b b
```
Ok :+1:

miguel-negrao commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -168,4 +196,75 @@ test = testGroup "Utility unit tests"
+-- | A type class for induction on the type-level list containing the Traversables. 
+class GenTests a where
+  genTraversableTests ::  
+    (forall (t :: * -> *) . (Arbitrary (t Int), Show (t Int), Eq (t Int), Traversable t, Typeable t) => Proxy t -> Test)
```
If I change all the occurences to Arbitrary1, Show1, Eq1, etc. And add

```Haskell
instance  (KnownNat n) => Arbitrary1 (V.Vector n) where
  liftArbitrary = V.replicateM
  liftShrink = V.mapM
instance (Arbitrary a, KnownNat n) => Arbitrary (V.Vector n a) where
  arbitrary = arbitrary1
  shrink = shrink1 
```

Then I get this error

```Haskell
test/Cell/Util.hs:231:9: error:
    • Could not deduce (Arbitrary (t Int))
        arising from a use of ‘makeTraversableTest’
      from the context: (Arbitrary1 t, Show1 t, Eq1 t, Traversable t,
                         Typeable t)
        bound by a type expected by the context:
                   forall (t :: * -> *).
                   (Arbitrary1 t, Show1 t, Eq1 t, Traversable t, Typeable t) =>
                   Proxy t -> Test
        at test/Cell/Util.hs:231:9-54
    • In the second argument of ‘($)’, namely
        ‘makeTraversableTest (traverseArrLaw @Int @Int)’
```
I don't entirely undestand how these type classes work (Show1, etc). Do you have an ideia of what need to do more ?

miguel-negrao replied:
I had another look at this, and I don't think that `Arbitrary1` will help much here, since there isn't an automatic instance for `Arbitrary (t Int)` even if there is an instance for `Arbitrary1 t`, and `Arbitrary Int`. Such an instance can be made for a specific t easilly (`arbitrary = arbitrary1`), but it is not made automatically it seems, as it would clash with other instances. So I think I would leave it as it is.

miguel-negrao commented:
Sorry for leaving this PR abandoned for so long :-). I've tried to address all the points of the previous review. There are still 2 unresolved conversations above, but I've pushed my current version of the code.

miguel-negrao commented:
Ok, I think I have addressed all the issues of the last review.

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -192,4 +212,98 @@ test =
           }
     ]
 
+type TestTraversables' = TestTraversables '[Maybe, [], V.Vector 10, Seq, Map Int]
```
Can you put everything related to this in a separate module `Cell.Util.Traversable`

turion replied:
Can we rename:

* `TestTraversables' -> TestTraversables`
* `TestTraversables -> Traversables`

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -192,4 +212,98 @@ test =
+traverseArrLaw ::
+  forall a b t.
+  Traversable t =>
+  Proxy t ->
```
Is it possible to do everything with `TypeApplications` instead of `Proxy`?

turion replied:
If not, no worries.

miguel-negrao replied:
No I don't think it is. I don't think you can do partial application with type application, which I think would be needed...

turion commented on essence-of-live-coding/test/Util.hs:
```diff
@@ -86,3 +86,20 @@ instance (Eq b, Show b) => Testable (CellSimulation a b) where
+   Check whether one cell behaves the same as another cell.
+-}
+data CellIdentitySimulation a b = CellIdentitySimulation
+  { cell1' :: Cell Identity a b
+  , cell2' :: Cell Identity a b
+  , input' :: [a]
+  }
```
```suggestion
  { cellL :: Cell Identity a b
  , cellR :: Cell Identity a b
  , jointInput :: [a]
  }
```

turion commented on essence-of-live-coding/test/Cell/Util.hs:
```diff
@@ -192,4 +212,98 @@ test =
+          }
+    , testProperty "traverse' by itself does not force the entire list (Cell)" $
+        CellSimulation
+          { cell = cellId >>> arr head
```
```suggestion
          { cell = toCell $ arr head
```
And the definition of `cellId` can be deleted.

miguel-negrao commented:
Thanks for the review ! Will do the changes and report back.

miguel-negrao commented:
Hopefully I have addressed the suggestions in the last review. As always, note that when you think this is ready to merge I can first squash que commits, but for now I'm leaving the individual commits to make it easier to see my latest changes.

miguel-negrao commented:
Hi, perhaps this is ready to merge ?

miguel-negrao commented:
Great ! It took a while, but is was worth discussing things and polishing the code ! Thanks for your reviews ! :-)
